### ISSUE TITLE
Suggestion: macOS bundle compliance

### ISSUE THREAD

**Issue by cfcurtis** [Nuitka/Nuitka#2906]
## Context
I have a cross-platform wxPython program that I have been bundling and distributing with PyInstaller for several years. I recently discovered Nuitka and found that on Windows it created a smaller .exe and did not cause Windows Defender to freak out. Perfect! The .app on macOS is considerably larger, but the startup time is much faster and I think that is more important.

On macOS, I finally caved and gave Apple money so that my end users don't have to jump through security hoops. The code signing and notarization process is not at all straightforward, and I found that the Nuitka-built bundle failed notarization repeatedly. I went as far as isolating the issue: wxPython was not a problem, but for some reason using `os.path` or `pathlib.Path` caused the notarization to be invalid. 

Just as I was putting the finishing touches on a [minimal example](https://github.com/cfcurtis/nuitka-wx-app), it suddenly started passing notarization. Absolutely no idea why, I can only surmise that something changed at the Apple end of the black box. I'm leaving this example on GitHub in case anyone wants a small wxPython example to build from.

In the process of doing all this, I found that Nuitka creates somewhat non-standard app bundles, and I'd like to make some suggestions to make the (poorly documented and infuriating) notarization process smoother.

## Suggestions to improve macOS app bundle compliance
- Place dynamic libraries in `Contents/Framework` instead of `Contents/MacOS` according to the policies outlined [here](https://developer.apple.com/documentation/bundleresources/placing_content_in_a_bundle). The present behaviour seems to be working, but it might be a good idea to follow the rules in case they change the notarization in the future.
- Make sure that no data files are present in `Contents/MacOS` or `Contents/Frameworks`. This is one of the poorly documented parts of Apple notarization that has an even more vague error message, but thanks to #2232 I eventually figured out what it was complaining about. My symlink workaround looks like this:
  ```bash
  for dir in "various" "data" "folders"; do 
    mv "app.app/Contents/MacOS/$dir" "app.app/Contents/Resources/"
    cd "app.app/Contents/MacOS/" # needs to be a relative symlink
    ln -s "../Resources/$dir" "."
    cd -
  done
  ```
  I then run the `codesign` tool again as follows:
  ```bash
  codesign --deep --force --options runtime --timestamp --sign "$APPLE_SIGN_IDENTITY" app.app
  ```
  This is all very hacky, but it seems to work. I'm not sure the best way to incorporate this kind of thing into Nuitka, but as it stands notarization will fail if any data files are included, and fail in an obscure way. An easy band-aid right now might be to warn the user if data files are copied to an app bundle and the `--macos-sign-notarization` flag is present.
- One small thing: rename the .app bundle in addition to the internal binary. Right now I'm just doing it manually after running Nuitka but it took me a while to realize that the `output-filename` parameter was only applying to the internal binary and not the resulting .app.

I apologize for the novel, and thank you for all the work on this program. Here's my answers to the template questions, even though this is no longer a bug report.

## Answers to template questions
- Nuitka Version:

  ```bash
  $ python -m nuitka --version
  2.3
  Commercial: None
  Python: 3.10.13 (main, Aug 24 2023, 12:59:26) [Clang 15.0.0 (clang-1500.0.40.1)]
  Flavor: Homebrew Python
  Executable: /Users/charlotte/Documents/my-repos/nuitka-wx/venv/bin/python
  OS: Darwin
  Arch: arm64
  macOSRelease: 14.5
  Version C compiler: /usr/bin/clang (clang 15.0.0).
  ```

- How did you install Nuitka and Python?
  `pip install nuitka` from a Homebrew venv

- The specific PyPI names and versions
  ```
  Nuitka==2.3
  ordered-set==4.1.0
  pillow==10.3.0
  six==1.16.0
  wxPython==4.2.1
  zstandard==0.22.0
  ```

- Also supply a Short, Self Contained, Correct, Example
  
  A minimal working example app can be found at https://github.com/cfcurtis/nuitka-wx-app. Since this is no longer a bug report for an actual error, I'm just providing this in case someone wants a sample of how to create a mac app bundle, code sign it, and submit for notorization.

- Provide in your issue the Nuitka options used
  ```bash
  python -m nuitka \
    --standalone \
    --macos-create-app-bundle \
    --macos-app-name="Hello Nuitka" \
    --company-name="Testing" \
    --macos-app-version="0.1" \
    --macos-signed-app-name="com.testing.hello-nuitka" \
    --macos-sign-identity=auto \
    --macos-sign-notarization \
    --assume-yes-for-downloads \
    --output-filename="hello_nuitka" \
    --macos-app-icon="resources/icon.icns" \
    --script-name=test_app.py
  ```

**Comment by kayhayen** [comment]
You convinced me that Nuitka should create the symlinks for data files. Since `__file__` is used for very different things, there is no good way to separate third-party package data files from code without modifying and investigating packages. Them being symlinks is hopefully OK, but I am pretty sure somebody is going to have code that breaks due to that. Of course, you can also embed files then it's not an issue, as Nuitka commercial allows you to do: https://nuitka.net/doc/commercial/protect-data-files.html

As for the size of the distribution, once we provide a proper self-compiled CPython for macOS, maybe that'll be less of an issue, but indeed, macOS is the worst right now. This is mostly about figuring out how to make one that is good for embedding and still loading extension modules. There, then, we could trim things a lot better, I believe. However, I do not have the time myself to provide that yet; I will, however, want to have a set of recommended self-compiles for Win32, Linux, and macOS for maximum portability.

**Comment by kayhayen** [comment]
Please somebody try and attack this, I do not currently have the time to do this myself, yet changing the placement of data files in Nuitka is a central thing, and volunteers should be able to find it and execute it.

**Comment by kayhayen** [comment]
Compiled code is necessarily larger than the bytecode, nothing to do about that. It can of course execute faster then and avoid some overheads.

**Comment by Areopagitics** [comment]
I was able to make it work with the following additions in line 233 of Standalone.py:

```
    import shutil
    from nuitka.Options import getMacOSAppName

    name = getMacOSAppName()
    # Define source and destination directories
    source_dir = name +".app/Contents/MacOS"
    destination_dir = name +".app/Contents/Resources"


    # Move all contents recursively
    for root, dirs, files in os.walk(source_dir, topdown=True):
        # Process directories first to maintain structure
        for dir_name in dirs:
            source_path = os.path.join(root, dir_name)
            if dir_name == name:  # Skip main executable
                continue
            destination_path = os.path.join(destination_dir, os.path.relpath(source_path, source_dir))
            
            shutil.move(source_path, destination_path)  # Move directory
            os.symlink(os.path.relpath(destination_path, root), source_path)  # Create relative symlink

        # Process files within directories
        for file_name in files:
            source_path = os.path.join(root, file_name)
            destination_path = os.path.join(destination_dir, os.path.relpath(source_path, source_dir))
            
            shutil.move(source_path, destination_path)  # Move file
            os.symlink(os.path.relpath(destination_path, root), source_path)  # Create relative symlink

```

**Comment by kayhayen** [comment]
I am preparing a change, where the copying to dist folder, does this transparently, also for entry points like extension modules, etc. and that takes more time. But thanks for confirming it properly signs, I didn't know that and would actually have to unearth it.

**Comment by kayhayen** [comment]
I kind of failed to implement this while doing the copy, placing, etc. of files, as it's too complex for a hotfix, so on factory, there is now code very much alike what you suggested, that moves things around right before signing (if it's an app bundle), and then it works.

**Comment by kayhayen** [comment]
Thanks for your report. This is fixed on the factory branch, which is a development version that is under rapid development. You can try it out by going here: https://nuitka.net/doc/factory.html 
	
Feedback on whether this is working is very welcome. Please do not share plans to do it; only confirm or deny that it is working.

**Comment by kayhayen** [comment]
Part of the 2.7.1 hotfix release I just made.

**Comment by BRaynerCo3** [comment]
@kayhayen, one small note mentioned in the original issue was:

> One small thing: rename the .app bundle in addition to the internal binary. Right now I'm just doing it manually after running Nuitka but it took me a while to realize that the output-filename parameter was only applying to the internal binary and not the resulting .app.

this does not appear fixed in the latest version of nuitka (2.7.13). Is there any plan to add this behaviour? many thanks!

**Comment by kayhayen** [comment]
Does renaming the `.app` folder manually break the signature, or do these names have to match?

**Comment by BRaynerCo3** [comment]
it does not appear to break the signature. I'm not an expert in Mac Bundle construction or signing - but running `codesign -vv` on a renamed `.app` folder will still report as valid in my testing.

Incidentally, the `Info.plist` in your generated `.app` does correctly list the `CFBundleDisplayName`, `CFBundleName` and `CFBundleIdentifier` as per the specified output file name parameter, so it appears to only be the `.app` folder name which needs changing.

**Comment by kayhayen** [comment]
I agree that have --output-filename only affect the name inside the created folder is perfectly consistent, but also suprising. I need to check what else influences said folder name and then add something, but --output-dir is going to be confusing there.

**Comment by kayhayen** [comment]
It's already there for  a while, just read help output.
